lib/hanami/action/params.rb

```diff
@@ -16,6 +16,62 @@ module Action
     class Params < BaseParams
       include Hanami::Validations::Form
 
+      # Params errors
+      #
+      # @since 1.1.0
+      class Errors < SimpleDelegator
+        # @since 1.1.0
+        # @api private
+        def initialize(errors = {})
+          super(errors)
+        end
+
+        # Add an error to the param validations
+        #
+        # This has a semantic similar to `Hash#dig` where you use a set of keys
+        # to get a nested value, here you use a set of keys to set a nested
+        # value.
+        #
+        # @param args [Array<Symbol, String>] an array of arguments: the last
+        #   one is the message to add (String), while the beginning of the array
+        #   is made of keys to reach the .
+        #
+        # @since 1.1.0
+        #
+        # @example
+        #   require "hanami/controller"
+        #
+        #   class MyAction
+        #     include Hanami::Action
+        #
+        #     params do
+        #       required(:book).schema do
+        #         required(:title).filled(:str?)
+        #       end
+        #     end
+        #
+        #     def call(params)
+        #       # 1. Don't try to save the record if the params aren't valid
+        #       return unless params.valid?
+        #
+        #       BookRepository.new.create(params[:book])
+        #     rescue Hanami::Model::UniqueConstraintViolationError
+        #       # 2. Add an error in case the record wasn't unique
+        #       params.errors.add(:book, :isbn, "is not unique")
+        #     end
+        #   end
+        def add(*args)
+          *keys, key, error = args
+
+          (if keys.empty?
```

Comment:
WDYT, can we move this if condition to a private method? (for better readability)

Reply:
Done: https://github.com/hanami/controller/pull/232/commits/92d2334eb5368acd79c10df5009b152dec02f0e8